README.md

# ideasbox

The sandbox for your features about a project, request your namespace :)

Give power to your users.

## Contributors

- [Alejandro Morales](http://alejandromorales.co.cc)
- [Kronos](http://github.com/alejandromg/kronos 'npm install kronos') 

## Licence
 
**MIT 2012**


History.md

n.n.n / 2012-03-11 
==================

  * [master] Client error handler
  * [master] Added elipsis method and better render of idea
  * [master] Fixed methods
  * [master] Fixed names in methods
  * [masteR] Fixed the Method not Allowed
  * [master] Added locall suport
  * [master] First working model
  * [master] Added index and more
  * [master] Added foundation and more godness
  * [master] jade support for tako :)
  * [master] Added dependencies
  * [kronos] Initial Commit

n.n.n / 2012-03-11 
==================

  * [master] Client error handler
  * [master] Added elipsis method and better render of idea
  * [master] Fixed methods
  * [master] Fixed names in methods
  * [masteR] Fixed the Method not Allowed
  * [master] Added locall suport
  * [master] First working model
  * [master] Added index and more
  * [master] Added foundation and more godness
  * [master] jade support for tako :)
  * [master] Added dependencies
  * [kronos] Initial Commit

n.n.n / 2012-03-10 
==================

  * [master] Added index and more
  * [master] Added foundation and more godness
  * [master] jade support for tako :)
  * [master] Added dependencies
  * [kronos] Initial Commit

n.n.n / 2012-03-09 
==================

  * [master] Added index and more
  * [master] Added foundation and more godness
  * [master] jade support for tako :)
  * [master] Added dependencies
  * [kronos] Initial Commit


lib/idea.js

var prettyTitle = function (title) {
  return title
  // change everything to lowercase
  .toLowerCase() 
  // trim leading and trailing spaces
  .replace(/^\s+|\s+$/g, "") 
  // change all spaces and underscores to a hyphen
  .replace(/[_|\s]+/g, "-") 
  // remove all non-cyrillic, non-numeric characters except the hyphen
  .replace(/[^a-z\u0400-\u04FF0-9-]+/g, "") 
  // replace multiple instances of the hyphen with a single instance
  .replace(/[-]+/g, "-") 
  .replace(/^-+|-+$/g, "")
  // trim leading and trailing hyphens
  .replace(/[-]+/g, "-")
  .substr(0,20);
}

function Idea(idea){
  var newIdea = {
    _rev: idea._rev || null,
    _id: idea._id || null,
    name: idea.name || 'Untitled',
    url: prettyTitle(idea.name || 'Untitled'),
    description: idea.description || 'Without description',
    author: {
      name:(idea.author.name && idea.author.name !== '') ? idea.author.name : 'anon',
      password:idea.author.password ?idea.author.password : null ,
      from:(idea.author.from && idea.author.from !== '')?idea.author.from : 'invalid'
    },
    'created_at': idea['created_at'] || new Date(),
    up: idea.up || 0,
    down:idea.down || 0,
    room: '#'+(idea.room?idea.room.replace('#',''):idea.room) || '#nodester'
  };
  if (newIdea._rev === null) delete newIdea._rev
  if (newIdea._id === null) delete newIdea._id
  if (newIdea.author.password === null) delete newIdea.author.password
  return newIdea;
}

module.exports = Idea;

public/javascripts/ideas.js
$(document).ready(function(){
  var loader = $('#loader')
    , form = $('#idea')
    , comment = $('#comment')
    , submit = $('#send')
    , description = $('[name="description"]')
    , createRoom = $('#create')
    , successIdea = '<p> You can see your idea at:<a href="<%= url%>",title="Go to"><%= url%></a></p>'
    , msg = $('#msg')
    ;
    $('[type="text"]').val('');
    description.val('');
    $('[rel~="facebox"]').facebox();
    window.oldAlert = window.alert;
    window.alert = function(str){
      jQuery.facebox(str);
    }
    createRoom.on('click', function(e){
      e.preventDefault()
      jQuery.facebox(newRoom);
      description.val('')
    })
    
    submit.on('click', function(e){
      e.preventDefault();
      loader.removeClass('off')
      msg.textContent = 'Posting...'
      form.addClass('off')
      var action ='new'
      if (window.location.pathname.search('/create')!==-1) action = 'create'
      var data = serialize(action);
      $.post('/'+data.room + '/'+action,data,function(resp,status){
        if (resp && status == 'success'){
          var self = resp;
          switch (resp.statusCode){
            case 200:
            case '200':
              var tpl = _.template(successIdea)
              comment.html(tpl({url:resp.url}));
              break;
            case '422':
            case 422:
              var msg = resp.msg ? resp.msg:'Idea name already Exists'
              comment.html('<div class="center"><h4> Error</h4><span> '+msg+'</span></div>')
              break;
            default:
              comment.html('<div class="center"><h4> 500 Internal Server Error</h4><span> Try again in a little</span></div>')
              break;
          }
        }
      });
    });
    function getValues (name){
      return $('[name="'+name+'"]')
    }
    window.serialize = function(action){
      var serial = {
          name: getValues('name').val()
        , description : getValues('description').val()
        , author:{
              name: getValues('user[name]').val()
            , from : getValues('user[from]').val()
            , email: getValues('user[email]').val()
          }
        , room : getValues('user[from]').val()
      }
      switch (action){
        case "create":
            serial.author.password =getValues('user[password]').val()
            return serial;
          break;
        default:
          return serial;
        break;
      }
    }
  var newRoom ='\
  <section id="form" class="five columns centered" style="margin-left:10px;margin-right:10px">\
      <img src="/assets/images/loading.gif" id="loader" class="off"><br>\
      <h4 id="msg">Add your awesome room  \
      </h4><br>\
      <form action="/ideasBoxs/create" method="POST" id="sendBox" class="nice" style="margin-left:0">\
        <input type="text" name="name" placeholder="the name of your room" required="required" class="input-text">\
        <textarea name="description" placeholder="tell us more about your room" cols="30" rows="5" required="required"></textarea>\
        <input type="text" name="user[name]" placeholder="your name" required="required" class="input-text">\
        <input type="text" name="user[email]" placeholder="your email (we won\'t store it) just for the avatar" required="required" class="input-text">\
        <input type="password" name="user[password]" placeholder="add your secret p4ssw0r2 (this is not recovered)" class="input-text">\
        <input type="hidden" name="user[from]" value="ideasBoxs">\
        <input type="submit" value="add new Room" id="send" class="button black radius">\
      </form>\
    </section>'
})

public/javascripts/facebox.js
/*
 * Facebox (for jQuery)
 * version: 1.2 (05/05/2008)
 * @requires jQuery v1.2 or later
 *
 * Examples at http://famspam.com/facebox/
 *
 * Licensed under the MIT:
 *   http://www.opensource.org/licenses/mit-license.php
 *
 * Copyright 2007, 2008 Chris Wanstrath [ chris@ozmm.org ]
 *
 * Usage:
 *
 *  jQuery(document).ready(function() {
 *    jQuery('a[rel*=facebox]').facebox()
 *  })
 *
 *  <a href="#terms" rel="facebox">Terms</a>
 *    Loads the #terms div in the box
 *
 *  <a href="terms.html" rel="facebox">Terms</a>
 *    Loads the terms.html page in the box
 *
 *  <a href="terms.png" rel="facebox">Terms</a>
 *    Loads the terms.png image in the box
 *
 *
 *  You can also use it programmatically:
 *
 *    jQuery.facebox('some html')
 *    jQuery.facebox('some html', 'my-groovy-style')
 *
 *  The above will open a facebox with "some html" as the content.
 *
 *    jQuery.facebox(function($) {
 *      $.get('blah.html', function(data) { $.facebox(data) })
 *    })
 *
 *  The above will show a loading screen before the passed function is called,
 *  allowing for a better ajaxy experience.
 *
 *  The facebox function can also display an ajax page, an image, or the contents of a div:
 *
 *    jQuery.facebox({ ajax: 'remote.html' })
 *    jQuery.facebox({ ajax: 'remote.html' }, 'my-groovy-style')
 *    jQuery.facebox({ image: 'stairs.jpg' })
 *    jQuery.facebox({ image: 'stairs.jpg' }, 'my-groovy-style')
 *    jQuery.facebox({ div: '#box' })
 *    jQuery.facebox({ div: '#box' }, 'my-groovy-style')
 *
 *  Want to close the facebox?  Trigger the 'close.facebox' document event:
 *
 *    jQuery(document).trigger('close.facebox')
 *
 *  Facebox also has a bunch of other hooks:
 *
 *    loading.facebox
 *    beforeReveal.facebox
 *    reveal.facebox (aliased as 'afterReveal.facebox')
 *    init.facebox
 *    afterClose.facebox
 *
 *  Simply bind a function to any of these hooks:
 *
 *   $(document).bind('reveal.facebox', function() { ...stuff to do after the facebox and contents are revealed... })
 *
 */
(function($) {
  $.facebox = function(data, klass) {
    $.facebox.loading()

    if (data.ajax) fillFaceboxFromAjax(data.ajax, klass)
    else if (data.image) fillFaceboxFromImage(data.image, klass)
    else if (data.div) fillFaceboxFromHref(data.div, klass)
    else if ($.isFunction(data)) data.call($)
    else $.facebox.reveal(data, klass)
  }

  /*
   * Public, $.facebox methods
   */

  $.extend($.facebox, {
    settings: {
      opacity      : 0.2,
      overlay      : true,
      loadingImage : '/assets/images/loading.gif',
      closeImage   : '/assets/images/closelabel.png',
      imageTypes   : [ 'png', 'jpg', 'jpeg', 'gif' ],
      faceboxHtml  : '\
    <div id="facebox" style="display:none;"> \
      <div class="popup"> \
        <div class="content"> \
        <header class="modal_header">\
        </header> \
        </div> \
        <a href="#" class="close"><img src="/assets/images/closelabel.png" title="close" class="close_image" /></a> \
      </div> \
    </div>'
    },

    loading: function() {
      init()
      if ($('#facebox .loading').length == 1) return true
      showOverlay()

      $('#facebox .content').empty()
      $('#facebox .body').children().hide().end().
        append('<div class="loading"><img src="'+$.facebox.settings.loadingImage+'"/></div>')

      $('#facebox').css({
        top:	getPageScroll()[1] + (getPageHeight() / 10),
        left:	$(window).width() / 2 - 205
      }).show()

      $(document).bind('keydown.facebox', function(e) {
        if (e.keyCode == 27) $.facebox.close()
        return true
      })
      $(document).trigger('loading.facebox')
    },

    reveal: function(data, klass,header) {
      $(document).trigger('beforeReveal.facebox')
      if (klass) $('#facebox .content').addClass(klass)
      $('#facebox .content').append(header || '<header class="header">ideasBox</header>')
      $('#facebox .content').append(data)
      $('#facebox .loading').remove()
      $('#facebox .body').children().fadeIn('normal')
      $('#facebox').css('left', $(window).width() / 2 - ($('#facebox .popup').width() / 2))
      $(document).trigger('reveal.facebox').trigger('afterReveal.facebox')
    },

    close: function() {
      $(document).trigger('close.facebox')
      return false
    }
  })

  /*
   * Public, $.fn methods
   */

  $.fn.facebox = function(settings) {
    if ($(this).length == 0) return

    init(settings)

    function clickHandler() {
      $.facebox.loading(true)

      // support for rel="facebox.inline_popup" syntax, to add a class
      // also supports deprecated "facebox[.inline_popup]" syntax
      var klass = this.rel.match(/facebox\[?\.(\w+)\]?/)
      if (klass) klass = klass[1]

      fillFaceboxFromHref(this.href, klass)
      return false
    }

    return this.bind('click.facebox', clickHandler)
  }

  /*
   * Private methods
   */

  // called one time to setup facebox on this page
  function init(settings) {
    if ($.facebox.settings.inited) return true
    else $.facebox.settings.inited = true

    $(document).trigger('init.facebox')
    makeCompatible()

    var imageTypes = $.facebox.settings.imageTypes.join('|')
    $.facebox.settings.imageTypesRegexp = new RegExp('\.(' + imageTypes + ')$', 'i')

    if (settings) $.extend($.facebox.settings, settings)
    $('body').append($.facebox.settings.faceboxHtml)

    var preload = [ new Image(), new Image() ]
    preload[0].src = $.facebox.settings.closeImage
    preload[1].src = $.facebox.settings.loadingImage

    $('#facebox').find('.b:first, .bl').each(function() {
      preload.push(new Image())
      preload.slice(-1).src = $(this).css('background-image').replace(/url\((.+)\)/, '$1')
    })

    $('#facebox .close').click($.facebox.close)
    $('#facebox .close_image').attr('src', $.facebox.settings.closeImage)
  }

  // getPageScroll() by quirksmode.com
  function getPageScroll() {
    var xScroll, yScroll;
    if (self.pageYOffset) {
      yScroll = self.pageYOffset;
      xScroll = self.pageXOffset;
    } else if (document.documentElement && document.documentElement.scrollTop) {	 // Explorer 6 Strict
      yScroll = document.documentElement.scrollTop;
      xScroll = document.documentElement.scrollLeft;
    } else if (document.body) {// all other Explorers
      yScroll = document.body.scrollTop;
      xScroll = document.body.scrollLeft;
    }
    return new Array(xScroll,yScroll)
  }

  // Adapted from getPageSize() by quirksmode.com
  function getPageHeight() {
    var windowHeight
    if (self.innerHeight) {	// all except Explorer
      windowHeight = self.innerHeight;
    } else if (document.documentElement && document.documentElement.clientHeight) { // Explorer 6 Strict Mode
      windowHeight = document.documentElement.clientHeight;
    } else if (document.body) { // other Explorers
      windowHeight = document.body.clientHeight;
    }
    return windowHeight
  }

  // Backwards compatibility
  function makeCompatible() {
    var $s = $.facebox.settings

    $s.loadingImage = $s.loading_image || $s.loadingImage
    $s.closeImage = $s.close_image || $s.closeImage
    $s.imageTypes = $s.image_types || $s.imageTypes
    $s.faceboxHtml = $s.facebox_html || $s.faceboxHtml
  }

  // Figures out what you want to display and displays it
  // formats are:
  //     div: #id
  //   image: blah.extension
  //    ajax: anything else
  function fillFaceboxFromHref(href, klass) {
    // div
    if (href.match(/#/)) {
      var url    = window.location.href.split('#')[0]
      var target = href.replace(url,'')
      if (target == '#') return
      $.facebox.reveal($(target).html(), klass)

    // image
    } else if (href.match($.facebox.settings.imageTypesRegexp)) {
      fillFaceboxFromImage(href, klass)
    // ajax
    } else {
      fillFaceboxFromAjax(href, klass)
    }
  }

  function fillFaceboxFromImage(href, klass) {
    var image = new Image()
    image.onload = function() {
      $.facebox.reveal('<div class="image"><img src="' + image.src + '" /></div>', klass)
    }
    image.src = href
  }

  function fillFaceboxFromAjax(href, klass) {
    $.get(href, function(data) { $.facebox.reveal(data, klass) })
  }

  function skipOverlay() {
    return $.facebox.settings.overlay == false || $.facebox.settings.opacity === null
  }

  function showOverlay() {
    if (skipOverlay()) return

    if ($('#facebox_overlay').length == 0)
      $("body").append('<div id="facebox_overlay" class="facebox_hide"></div>')

    $('#facebox_overlay').hide().addClass("facebox_overlayBG")
      .css('opacity', $.facebox.settings.opacity)
      .click(function() { $(document).trigger('close.facebox') })
      .fadeIn(200)
    return false
  }

  function hideOverlay() {
    if (skipOverlay()) return

    $('#facebox_overlay').fadeOut(200, function(){
      $("#facebox_overlay").removeClass("facebox_overlayBG")
      $("#facebox_overlay").addClass("facebox_hide")
      $("#facebox_overlay").remove()
    })

    return false
  }

  /*
   * Bindings
   */

  $(document).bind('close.facebox', function() {
    $(document).unbind('keydown.facebox')
    $('#facebox').fadeOut(function() {
      $('#facebox .content').removeClass().addClass('content')
      $('#facebox .loading').remove()
      $(document).trigger('afterClose.facebox')
    })
    hideOverlay()
  })

})(jQuery);


lib/ideasbox.js
#!/usr/bin/env node

/*
 * ideasbox
 * @date:Thu Mar 08 2012 21:25:59 GMT-0600 (CST)
 * @name: ideasbox.js
 * @licence: MIT
*/


var http    = require('http')
  , tako    = require('tako')
  , request = require('request')
  , config  = require('./config')
  , nano    = require('nano')(config.bd)
  , db      = nano.use('ideas')
  , app     = tako()
  , path    = require('path')
  , jade    = require('jade')
  , fs      = require('fs')
  , Idea    = require('./idea')
  , qs      = require('querystring')
  , url     = require('url')
  , Emitter  = require('eventemitter2').EventEmitter2;
  
var iB = new Emitter({
  namespace: 'ideasBox',
  delimiter: '::',
  maxListeners: 20,
  wildcard:  '*'
});

function JadeCompiler(home,globals){
  this.HOME   = path.join(home);
  this.layout = path.join(this.HOME, globals.layout || 'layout.jade');
  this.GLOBAL = globals;
}
JadeCompiler.prototype.render = function(file, locals){
  if (path.extname(file) === '') file += '.jade'
  if (path.extname(file) === '.jade') {
    var layout  = locals.layout ? path.join(this.HOME,locals.layout) : this.layout
      , _layout = fs.readFileSync(layout,'utf8')
      , _file   = fs.readFileSync(path.join(this.HOME, file), 'utf8')
      , page    = jade.compile(_file , { pretty : true })
      ;
    // At the end GLOBAL is not that "GLOBAL" as we want 
    for (var i in locals){
      this.GLOBAL[i] = locals[i]
    }
    page = page(this.GLOBAL);
    this.GLOBAL['body'] = page;
    var pageRenderized = jade.compile(_layout , { pretty : true });
    pageRenderized = pageRenderized(this.GLOBAL);
    return pageRenderized
  } else {
    return new Error('Invalid Extension Name: ' + path.extname(file))
  }
}
var views = new JadeCompiler(path.join(__dirname,'..','views'),{
  sitename:'ideasBox',
  author:'Alejandro Morales'
});

// Static Provider
app.route('/assets/*').files(path.join(__dirname,'..','public'));

app.route('/').html(function(req,res){
  iB.emit('stats::home',req);
  res.end(fs.readFileSync(__dirname+'/../public/untitled.html', 'utf8'))
}).methods('GET')

app.route('/favicon.ico', function(req,res){
  res.redirect('/assets/logo.png')
})
iB.on('stats::*',function(data){
  console.log(data.url)
})
iB.on('ideas::*',function(idea){
  
});
app.route('/home',function(req,res){
  res.redirect('/assets/images/misc/input-bg-outset.png')
})
app.route('/404', function(req,res){
  
  req.query= qs.parse(req.query)
  res.statusCode = 404;
  res.end(views.render('error',{
      code    : 404
    , info    : 'Page doesn\'t exists'
    , referer : unescape(req.query.ref)
  }))
});
app.route('/500', function(req,res){
  res.statusCode = 500;
  res.end(views.render('error',{
      code : 500
    , info : 'Internal Server Error'
  }))
})

var queryfied = function (data){
  data = new Buffer(data);
  data = qs.parse(data.toString('utf8'));
  var original = Object.create(data)
  for (var member in data){
    if (member.search('\\[') != -1){
      var key = member.split('[');
      data[key[0]] = data[key[0]]||{}
      data[key[0]][key[1].replace(']','')] = data[member]
      delete data[member]
    } else {
    }
  }
  data.name = original.name
  return data;
};

var saveRoom = function(req,res,data){
  data.user ? data.author = data.user : true;
  data = new Idea(data);
  data.channel = true;
  var room = data.room = '#'+req.route.params.room;
  var referer = url.parse(req.headers.referer)
  db.view('sort','channel',{key:unescape(room)},function(e,d){
    if (!e && d.rows.length > 0 ) {
      res.end({
          statusCode:422
        , msg: 'Room already exists'
      })
    } else {
      db.insert(data, function(e,d){
        var statusCode = 500
          , url        ='/';
        if (!e) {
          statusCode = 200
          url        = '/'+data.room.replace('#','')+'/'
        } 
        if (referer.pathname.search('create')!=-1){
          res.redirect(url);
        } else{
          res.end({
            statusCode : statusCode
          , url        : url
          })
        }
      });
    }
  });
};
app.route('/:room', function(req,res){
  var room = req.route.params.room;
      room = '#'+room
      room = escape(room)
  {"/_design/sort/_view/channel"}
  db.view('sort','channel',{key:unescape(room)},function(e,d){
    
     if (!e && d.rows.length > 0 ) {
       res.end(views.render('page', {
           ideas : d.rows
         , room  : req.route.params.room
       }));
     } else {
       res.redirect('/404?ref='+escape(room));
     }
  });
}).methods('GET')

app.route('/:room/new', function(req,res){
  req.on('data',function(da){
    
    var data = queryfied(da);
    if (data && !data.author) {
      data.author = {
          name :'Anon'
        , from: data.user
        , email :'aslda@node.com'
      }
    }

    var user = {
        name  : data.author.name
      , from  : data.author.from
      , email : data.author.email
    }
    var idea = {
        name        : data.name
      , description : data.description
      , author      : user
      , room        : data.room
    }
    var doc = new Idea(idea);
    var referer = url.parse(req.headers.referer)
    //based on default data not the doc, validation
    if (data.name.trim() ==='' || data.description.trim()===''  || /^-?(create|new|put|delete|patch|edit)*@/.test('-/'+data.name.trim()+'@') || doc.description.length < 200){
      var msg = 'The data is incorrect, invalid or too short';
      var code = 422;
      if (referer.pathname!=='/') {
        res.end({
            statusCode:code
          , info: 'Invalid Data'
          , msg: msg
        })
      } else {
        res.end(views.render('error',{
            code: code
          , info :msg
          , msg: msg
        }))
      }
    } else {
      db.view('sort','post',{key:[doc.name,'#'+doc.room]},function(e,d){
        if (!e && d.rows.length > 0) {
          res.statusCode = 422;
          res.end({
              statusCode : 422
            , info       : 'Identity not valid'
            , msg        : doc.name+': already exists.'
          })
        } else if (e){
          res.end({
            statusCode:500
          })
        } else {
          db.insert(doc,function(e,d){
            var statusCode = 500
              , url        ='/';
            if (!e) {
              statusCode = 200
              url        = '/'+doc.room.replace('#','')+'/'+doc.url
            } 
            
            if (referer.pathname!=='/'){
              res.end({
                  statusCode : statusCode
                , url        : '/'+doc.room.replace('#','')+'/'+doc.url
              })
            } else {
              switch (statusCode){
                case 500:
                  res.redirect('/500')
                default:
                  res.redirect(url);
                  break;
              }
            }
          })
        }
      });
    }
  })
}).methods('POST');


app.route('/:room/create',function(req,res){
  with (req.route.params){
    var room = room; // ? weird eh?
    function get(){
      var _room = '#'+ req.route.params.room;
      db.view('sort','channel',{key:unescape(_room)},function(e,d){
        if (!e && d.rows.length > 0) {
          res.redirect('/'+req.route.params.room);
        } else {
          res.end(views.render('create',{
            room: req.route.params.room
          }))
        }
      });
    }
    function post(){
      req.on('data',function(da){
        var data = queryfied(da);
        saveRoom(req,res,data);
      });
    }
    switch (req.method) {
      case 'POST':
        post();
        break;
      case 'GET':
        get()
        break;
      default:
        res.end('NOT ALLOWED');
    }
  }
}).methods('GET','POST')



app.route('/:room/:idea', function(req,res){
  with (req.route.params) {
    db.view('sort','url',{ key : [idea, '#'+room] , limit:1 }, function(e,d){
      if (!e && d.rows.length > 0){
        res.end(views.render('single',{
            room : room
          , idea : d.rows[0].value
        }))
      } else {
        res.redirect('/404')
      }
    });
  }
}).methods('GET')

var idea = {
    name:'Prueba con rinocerontes'
  , description:'Prueba con rinocerontes\n debe de estar prohibido ya que es inmoral y padece de interes particular de las personas por mejroar lo que se tiene.\nPrueba con rinocerontes\n debe de estar prohibido ya que es inmoral y padece de interes particular de las personas por mejroar lo que se tiene.  '
  , author:{
       name:'Alejandro Morales'
     , from:'ideasBox'
     , email:'vamg008@gmail.com'
    }
  , room:'ideasBox'
}
/*db.insert(new Idea(idea), function(e,d){
  console.log(e,d)
})*/




app.sockets.on('connection', function (socket) {
  app.sockets.emit('news', { will: 'be received by everyone'});
  socket.on('disconnect', function () {
    app.sockets.emit('user disconnected')
  })
})
function run(port) {
  port = port || 8000;
  app.httpServer.listen(port, function() {
    log('info','running on port => '+port,'ideasBox');
  });
};
module.exports.run = run;


var colors = {
  'bold'      : ['\033[1m',  '\033[22m'],
  'italic'    : ['\033[3m',  '\033[23m'],
  'underline' : ['\033[4m',  '\033[24m'],
  'inverse'   : ['\033[7m',  '\033[27m'],
  'white'     : ['\033[37m', '\033[39m'],
  'grey'      : ['\033[90m', '\033[39m'],
  'black'     : ['\033[30m', '\033[39m'],
  'blue'      : ['\033[34m', '\033[39m'],
  'cyan'      : ['\033[36m', '\033[39m'],
  'green'     : ['\033[32m', '\033[39m'],
  'magenta'   : ['\033[35m', '\033[39m'],
  'red'       : ['\033[31m', '\033[39m'],
  'yellow'    : ['\033[33m', '\033[39m']
}

var log = function(level, log,title){
  var color =[]
  switch (level){
    case 'info':
      color = colors['cyan']
      break;
    case 'log':
      color = colors['magenta']
      break;
    case 'warn':
      color = colors['red']
      break;
    default:
      color = color['green']
     break;
  }
  console[level](colors['bold'][0],(title?title:'DEBUG'),':',colors['bold'][1],color[0],log,color[1]);
}